mailheader.js

```diff
@@ -99,57 +99,77 @@ function _decode_header (matched, encoding, lang, cte, data) {
     return data.toString();
 }
 
-function _decode_rfc2231 (params) {
-    return function (matched, str) {
-        var sub_matches = /^(([^=]*)\*)(\d*)=(\s*".*?[^\\]";?|\S*)\s*$/.exec(str);
-        if (!sub_matches) {
-            return " " + str;
-        }
-        var key = sub_matches[1];
-        var key_actual = sub_matches[2];
-        var key_id = sub_matches[3] || '0';
-        var value = sub_matches[4].replace(/;$/, '');
-
-        var key_extract = /^(.*?)(\*(\d+)\*)$/.exec(key);
-        if (key_extract) {
-            key_actual = key_extract[1];
-            key_id = key_extract[3];
-        }
+function _decode_rfc2231 (params, str) {
+    /*
+    To explain the regexp below, the params are:
+
+    parameter := attribute "=" value
+
+    attribute := token
+                 ; Matching of attributes
+                 ; is ALWAYS case-insensitive.
+
+    token := 1*<any (US-ASCII) CHAR except SPACE, CTLs,
+                or tspecials>
+
+    tspecials :=  "(" / ")" / "<" / ">" / "@" /
+                  "," / ";" / ":" / "\" / <">
+                  "/" / "[" / "]" / "?" / "="
+                  ; Must be in quoted-string,
+                  ; to use within parameter values
+    */
+    var sub_matches = /(([!#$%&'*+.0-9A-Zdiff^_`a-z{|}~-]*)\*)(\d*)=(\s*".*?[^\\]";?|\S*)/.exec(str);
+    if (!sub_matches) {
+        return;
+    }
+    var key = sub_matches[1];
+    var key_actual = sub_matches[2];
+    var key_id = sub_matches[3] || '0';
+    var value = sub_matches[4].replace(/;$/, '');
```

Comment:
Is it too soon to use es6 destructing assignment? IE:

```js
let [, key, key_actual, key_id, value] = /(([!#$%&'*+.0-9A-Zdiff^_`a-z{|}~-]*)\*)(\d*)=(\s*".*?[^\\]";?|\S*)/.exec(str);
if (!key) return;
value = value.replace(/;$/, '');
```

Reply:
I'm not even sure this works on regexps because it's not returning an array, it's some funky object which happens to have 0,1,2,3, etc, but also has .index and .input values.

Anyway, I'm not going to change it. It works, and that trumps my time for experimenting ;)

Reply:
I haven't used it for regex yet myself but you around my curiosity. It seems to work as advertised:
```js
$ node
> let [,a,b]= /(the) (rain) (in)/.exec('the rain in spain');
undefined
> console.log(a);
the
undefined
```

We'll update it later when we get around to #2075